### PR TITLE
redesign the access control mechanism

### PR DESCRIPTION
**Pull request by Baalmart**
**_WHAT DOES THIS PR DO?_**

- [x] Creation of the appropriate endpoints for the access control mechanism of the analytics platform. With a focus on mainly four (4) entities which are: `Users`, `Networks`, `Permissions` and `Access Tokens`
- [x] Refactor the entire microservice to meet best practices and improve performance where possible. Apply rate limiting, reorganise files, etc.
- [x] Resolve a number of runtime errors in the microsevice.
- [ ] Migrate KCCA users data to the AirQo space
- [x] Adding scope and expiry dates to user tokens
- [ ] Allow guest user Login
- [ ] Address JWT security concerns where it is easy to decode the token without the secret in the browser.

**_WHAT ISSUES ARE RELATED TO THIS PR?_**

- Jira cards
    - [AN-148]

**_HOW DO I TEST OUT THIS PR?_**
```
cd src/auth-service
npm install
```
`npm run dev-mac` / `npm run dev-pc`
Or just use the deploy previews.

**_WHICH ENDPOINTS SHOULD BE READY FOR TESTING?:_** 

### Networks:
- [x] [create network](https://airqo-engineering.postman.co/workspace/internal~7a8e9fad-2d13-4d27-af27-f1f4e9a114bf/request/12513372-4474e6ea-5d1f-41b5-845b-1e409dff6908)
- [x] [Get network list](https://airqo-engineering.postman.co/workspace/internal~7a8e9fad-2d13-4d27-af27-f1f4e9a114bf/request/12513372-d392e03e-7372-4485-88fe-f844304a7b78)
- [x] [Get network information](https://airqo-engineering.postman.co/workspace/internal~7a8e9fad-2d13-4d27-af27-f1f4e9a114bf/request/12513372-3380a8e5-c840-47bc-807e-f592b40c9745)
- [x] [update networks](https://airqo-engineering.postman.co/workspace/internal~7a8e9fad-2d13-4d27-af27-f1f4e9a114bf/request/12513372-648d4944-d25b-447b-9c3f-2c5df16c91da)
- [x] [delete network](https://airqo-engineering.postman.co/workspace/internal~7a8e9fad-2d13-4d27-af27-f1f4e9a114bf/request/12513372-4e0f0c70-60f8-4f0c-934f-c3417f406476)
- [x] [Get users list for network.](https://airqo-engineering.postman.co/workspace/internal~7a8e9fad-2d13-4d27-af27-f1f4e9a114bf/request/12513372-f12b10b7-5623-4f40-8cb6-8754279a9e8f)
- [x] [Available Users List for Network](https://airqo-engineering.postman.co/workspace/internal~7a8e9fad-2d13-4d27-af27-f1f4e9a114bf/request/12513372-314142d6-d6f0-4e29-a04c-3906276ab09c)
- [x] [v1 Assign User to Network](https://airqo-engineering.postman.co/workspace/internal~7a8e9fad-2d13-4d27-af27-f1f4e9a114bf/request/12513372-af3835bc-0f92-4ec6-bc1f-5517670e6615)
- [x] [Assign User to Network](https://airqo-engineering.postman.co/workspace/internal~7a8e9fad-2d13-4d27-af27-f1f4e9a114bf/request/12513372-23d6bb59-bf3c-47d7-bba9-976b84b36216)
- [x] [UnAssign User from Network](https://airqo-engineering.postman.co/workspace/internal~7a8e9fad-2d13-4d27-af27-f1f4e9a114bf/request/12513372-70f6e218-20d8-4cd4-8fc3-1eadd4b3c890)
- [x] [Set manager for Network](https://airqo-engineering.postman.co/workspace/internal~7a8e9fad-2d13-4d27-af27-f1f4e9a114bf/request/12513372-95126b4d-d264-4593-9fd9-dce5a836fca0)
- [x] [check if network exists using email](https://airqo-engineering.postman.co/workspace/internal~7a8e9fad-2d13-4d27-af27-f1f4e9a114bf/request/12513372-1d2061fc-8151-4f5c-8762-534dc2265fcd)

### Users:

- [x] [Get Users List: GET /users](https://airqo-engineering.postman.co/workspace/internal~7a8e9fad-2d13-4d27-af27-f1f4e9a114bf/request/12513372-21a7b9e6-8204-4d21-a6b2-292c59a09025)
- [x] [Get User Information: GET /user/{usr_uid}](https://airqo-engineering.postman.co/workspace/internal~7a8e9fad-2d13-4d27-af27-f1f4e9a114bf/request/12513372-a51b127c-d20c-46ca-9ca5-e927d04ddd82)
- [x] [Create User: POST /user](https://airqo-engineering.postman.co/workspace/internal~7a8e9fad-2d13-4d27-af27-f1f4e9a114bf/request/12513372-8cfe27b0-0d50-4c72-b809-b728172d9716)
- [ ] Upload User Image: POST /user/{usr_uid}/image-upload
- [x] [Update User Information: PUT /user/{usr_uid}](https://airqo-engineering.postman.co/workspace/internal~7a8e9fad-2d13-4d27-af27-f1f4e9a114bf/request/12513372-f2581c1d-7182-4f2c-84d8-cc116873e23a)
- [x] [Delete User: DELETE /user/{usr_uid}](https://airqo-engineering.postman.co/workspace/internal~7a8e9fad-2d13-4d27-af27-f1f4e9a114bf/request/12513372-c5b7b9bb-156b-4b82-bada-bd843205d154)
- [x] [v1.1 Create User](https://airqo-engineering.postman.co/workspace/internal~7a8e9fad-2d13-4d27-af27-f1f4e9a114bf/request/12513372-4d34aaee-0d8c-442e-8782-bedc8e28d7e0)
- [x] [verify email link](https://airqo-engineering.postman.co/workspace/internal~7a8e9fad-2d13-4d27-af27-f1f4e9a114bf/request/12513372-0eaa3d72-e33e-484b-854c-a9a8940db741)
- [x] [Guest Login](https://airqo-engineering.postman.co/workspace/internal~7a8e9fad-2d13-4d27-af27-f1f4e9a114bf/request/12513372-d42b8817-37ac-4660-a82c-031b59679480)

### Clients

- [x] [Create Client](https://airqo-engineering.postman.co/workspace/internal~7a8e9fad-2d13-4d27-af27-f1f4e9a114bf/request/12513372-7a4bd396-cf14-4b22-9a34-98a8a3014c78)
- [x] [Get Clients](https://airqo-engineering.postman.co/workspace/internal~7a8e9fad-2d13-4d27-af27-f1f4e9a114bf/request/12513372-552a3d71-8b30-40b0-b24f-8fbc4a094d69)
- [x] [Get Client Information](https://airqo-engineering.postman.co/workspace/internal~7a8e9fad-2d13-4d27-af27-f1f4e9a114bf/request/12513372-7b109d79-1ae8-4950-aad8-7e7f24750099)
- [x] [Delete Client](https://airqo-engineering.postman.co/workspace/internal~7a8e9fad-2d13-4d27-af27-f1f4e9a114bf/request/12513372-4691787e-7aae-49ff-8e79-8b35f92328f9)
- [x] [Update Client Details](https://airqo-engineering.postman.co/workspace/internal~7a8e9fad-2d13-4d27-af27-f1f4e9a114bf/request/12513372-0ddeb0d8-a12e-4888-ba3b-af1b238085ef)
- [x] [Regenerate Client ID](https://airqo-engineering.postman.co/workspace/internal~7a8e9fad-2d13-4d27-af27-f1f4e9a114bf/request/12513372-ec79b608-6a31-4af1-87d8-24e0c60106eb)
- [x] [Update Client Name](https://airqo-engineering.postman.co/workspace/internal~7a8e9fad-2d13-4d27-af27-f1f4e9a114bf/request/12513372-2e68f910-bb8a-4a90-9764-883323f54af3)
- [x] [Regenerate Client Secret](https://airqo-engineering.postman.co/workspace/internal~7a8e9fad-2d13-4d27-af27-f1f4e9a114bf/request/12513372-16be2aff-ff93-426a-879a-02c17a6507ed)

### Permissions:

- [x] [Get Permissions List: GET /permissions](https://airqo-engineering.postman.co/workspace/internal~7a8e9fad-2d13-4d27-af27-f1f4e9a114bf/request/12513372-27d15c1e-7461-48be-bbd8-efc7c8e66c76)
- [x] [Get Permission Information: GET /permissions/{perm_id}](https://airqo-engineering.postman.co/workspace/internal~7a8e9fad-2d13-4d27-af27-f1f4e9a114bf/request/12513372-9f174e11-09f8-4070-8fa2-483c00e86d57)
- [x] [Create Permission: POST /permissions](https://airqo-engineering.postman.co/workspace/internal~7a8e9fad-2d13-4d27-af27-f1f4e9a114bf/request/12513372-fc8c024a-77f3-40dc-8c15-c0d726221eca)
- [x] [Update Permission: PUT /permissions/{perm_id}](https://airqo-engineering.postman.co/workspace/internal~7a8e9fad-2d13-4d27-af27-f1f4e9a114bf/request/12513372-2d1d0089-05f9-4c64-8830-d5e94e285c06)
- [x] [Delete Permission: DELETE /permissions/{perm_id}](https://airqo-engineering.postman.co/workspace/internal~7a8e9fad-2d13-4d27-af27-f1f4e9a114bf/request/12513372-d1e8ced4-e276-484f-bffa-9b8c73dc73da)

### Scopes

- [x] [Create Scope](https://airqo-engineering.postman.co/workspace/internal~7a8e9fad-2d13-4d27-af27-f1f4e9a114bf/request/12513372-748ab24d-7ce4-44ee-b5f2-873e15521878)
- [x] [Get Scopes](https://airqo-engineering.postman.co/workspace/internal~7a8e9fad-2d13-4d27-af27-f1f4e9a114bf/request/12513372-f6cdf3c8-711d-48c6-99f1-ed144d01efb4)
- [x] [Get Scope Information](https://airqo-engineering.postman.co/workspace/internal~7a8e9fad-2d13-4d27-af27-f1f4e9a114bf/request/12513372-7b60ba4e-03ef-4860-9a9a-fd3b14baa6d0)
- [x] [Delete Scope](https://airqo-engineering.postman.co/workspace/internal~7a8e9fad-2d13-4d27-af27-f1f4e9a114bf/request/12513372-24fb94db-e892-4eb5-9a83-dad2ce03af9e)
- [x] [Update Scope Details](https://airqo-engineering.postman.co/workspace/internal~7a8e9fad-2d13-4d27-af27-f1f4e9a114bf/request/12513372-2559ec10-81ce-445a-8a87-cc6fffbafe09)

### Access Tokens:

- [x] [Get Access Token List: GET /tokens](https://airqo-engineering.postman.co/workspace/internal~7a8e9fad-2d13-4d27-af27-f1f4e9a114bf/request/12513372-8c20dc88-ee19-4943-9ab7-abae371f7221)
- [x] [Get Access Token Information: GET /tokens/{token_id}](https://airqo-engineering.postman.co/workspace/internal~7a8e9fad-2d13-4d27-af27-f1f4e9a114bf/request/12513372-27f2da76-f59f-4133-97b4-1814609f49ca)
- [x] [Create Access Token: POST /tokens](https://airqo-engineering.postman.co/workspace/internal~7a8e9fad-2d13-4d27-af27-f1f4e9a114bf/request/12513372-7d50079a-1171-4ee2-8198-d19082edeb1d)
- [x] [Update Access Token: PUT /tokens/{token_id}](https://airqo-engineering.postman.co/workspace/internal~7a8e9fad-2d13-4d27-af27-f1f4e9a114bf/request/12513372-c29e976a-aa1f-4b2d-a8d9-ee3d7dda5e09)
- [x] [Delete Access Token: DELETE /tokens/{token_id}](https://airqo-engineering.postman.co/workspace/internal~7a8e9fad-2d13-4d27-af27-f1f4e9a114bf/request/12513372-bc2d12fb-01bf-44b3-9619-db67f380cfed)
- [x] [Verify Token](https://airqo-engineering.postman.co/workspace/internal~7a8e9fad-2d13-4d27-af27-f1f4e9a114bf/request/12513372-adf5f062-ac4d-4a33-a0cb-9167ca684a0b)

### Roles:

- [x] [Get Roles List: GET /roles](https://airqo-engineering.postman.co/workspace/internal~7a8e9fad-2d13-4d27-af27-f1f4e9a114bf/request/12513372-c964a0a6-eab8-4499-bb06-03fe2792bc79)
- [x] [Get Role Information: GET /role/{rol_uid}](https://airqo-engineering.postman.co/workspace/internal~7a8e9fad-2d13-4d27-af27-f1f4e9a114bf/request/12513372-4712237c-4a4b-4244-9412-b52fa15c2058)
- [x] [Create Role: POST /role](https://airqo-engineering.postman.co/workspace/internal~7a8e9fad-2d13-4d27-af27-f1f4e9a114bf/request/12513372-0168bd9b-095d-449b-9a10-0ed94debcab4)
- [x] [Update Role: PUT /role/{rol_uid}](https://airqo-engineering.postman.co/workspace/internal~7a8e9fad-2d13-4d27-af27-f1f4e9a114bf/request/12513372-71aceff7-e2e0-4315-b918-e864b8491e39)
- [x] [Delete Role: DELETE /role/{rol_uid}](https://airqo-engineering.postman.co/workspace/internal~7a8e9fad-2d13-4d27-af27-f1f4e9a114bf/request/12513372-5f745cc5-f138-48e0-881b-7679fc127d7f)
- [x] [Get Users with Role: GET /role/{rol_uid}/users](https://airqo-engineering.postman.co/workspace/internal~7a8e9fad-2d13-4d27-af27-f1f4e9a114bf/request/12513372-fa0a4f14-2e40-498e-8ab7-272dcdae4bc9)
- [x] [Available Users for Role: GET /role/{rol_uid}/available-users](https://airqo-engineering.postman.co/workspace/internal~7a8e9fad-2d13-4d27-af27-f1f4e9a114bf/request/12513372-b3927a42-7b6a-440d-97a2-5af53cd50882)
- [x] [Assign User to Role: POST /role/{rol_uid}/user](https://airqo-engineering.postman.co/workspace/internal~7a8e9fad-2d13-4d27-af27-f1f4e9a114bf/request/12513372-75adee8a-69c3-4f78-b964-c03ff47e1c64)
- [x] [Unassign User from Role: DELETE /role/{rol_uid}/user/{usr_uid}](https://airqo-engineering.postman.co/workspace/internal~7a8e9fad-2d13-4d27-af27-f1f4e9a114bf/request/12513372-657d6f47-0a1a-4385-9c61-6f8165e21355)
- [x] [Permissions List for Role: GET /role/{rol_uid}/permissions](https://airqo-engineering.postman.co/workspace/internal~7a8e9fad-2d13-4d27-af27-f1f4e9a114bf/request/12513372-7734f4e0-b924-4d5a-9327-5eebd63bff7e)
- [x] [Available Permissions for Role: GET /role/{rol_uid}/available-permissions](https://airqo-engineering.postman.co/workspace/internal~7a8e9fad-2d13-4d27-af27-f1f4e9a114bf/request/12513372-2f8d30ae-8c34-4fe7-9b30-01ea2da30641)
- [x] [Assign Permission to Role: POST /role/{rol_uid}/permissions](https://airqo-engineering.postman.co/workspace/internal~7a8e9fad-2d13-4d27-af27-f1f4e9a114bf/request/12513372-f96f2903-bd60-4df0-99bf-364be7427340)
- [x] [Unassign Permission from Role: DELETE /role/{rol_uid}/permissions/{per_uid}](https://airqo-engineering.postman.co/workspace/internal~7a8e9fad-2d13-4d27-af27-f1f4e9a114bf/request/12513372-ef7b6168-b611-44dc-b9cb-5194b16ce968)
- [ ] Create Roles from existing role
- [ ] Disable role

### Departments:

- [x] [Get Departments List: GET /departments](https://airqo-engineering.postman.co/workspace/internal~7a8e9fad-2d13-4d27-af27-f1f4e9a114bf/request/12513372-9969d049-4093-47c2-befc-0405c2d146a9)
- [ ] [Get Department Information: GET /departments/{dep_uid}](https://airqo-engineering.postman.co/workspace/internal~7a8e9fad-2d13-4d27-af27-f1f4e9a114bf/request/12513372-a4a67f78-c9a6-4053-abca-87ffccc7903c)
- [ ] [Create Department: POST /departments](https://airqo-engineering.postman.co/workspace/internal~7a8e9fad-2d13-4d27-af27-f1f4e9a114bf/request/12513372-31d94227-358b-42b7-af97-624352e0e2b4)
- [ ] [Update Department: PUT /departments/{dep_uid}](https://airqo-engineering.postman.co/workspace/internal~7a8e9fad-2d13-4d27-af27-f1f4e9a114bf/request/12513372-afe4847f-317b-4744-8e24-dd71250bceb1)
- [ ]  [Delete Department: DELETE /departments/{dep_uid}](https://airqo-engineering.postman.co/workspace/internal~7a8e9fad-2d13-4d27-af27-f1f4e9a114bf/request/12513372-c1f575c3-2774-4a55-a4de-b2e787360047)
- [ ] [Get Users List for Department: GET /departments/{dep_uid}/assigned-user](https://airqo-engineering.postman.co/workspace/internal~7a8e9fad-2d13-4d27-af27-f1f4e9a114bf/request/12513372-9da19507-f5a5-454b-9665-e9910236da4f)
- [ ] [Available Users List for Department: GET /departments/{dep_uid}/available-user](https://airqo-engineering.postman.co/workspace/internal~7a8e9fad-2d13-4d27-af27-f1f4e9a114bf/request/12513372-92367c84-4297-4665-b48c-a20bd6c9ffc7)
- [ ] [v1: Assign User to Department: PUT /departments/{dep_uid}/assign-user/{usr_uid}](https://airqo-engineering.postman.co/workspace/internal~7a8e9fad-2d13-4d27-af27-f1f4e9a114bf/request/12513372-34b3eeaf-3c97-48a3-8410-3c96194753e3)
- [ ] [Assign User to Department: POST /departments/{dep_uid}/assign-user](https://airqo-engineering.postman.co/workspace/internal~7a8e9fad-2d13-4d27-af27-f1f4e9a114bf/request/12513372-efa60201-8e56-4f51-b5dc-32e7a17113c1)
- [ ] [Unassign User from Department: DELETE /departments/{dep_uid}/unassign-user/{usr_uid}](https://airqo-engineering.postman.co/workspace/internal~7a8e9fad-2d13-4d27-af27-f1f4e9a114bf/request/12513372-7fa1a8df-0848-4ebb-9b61-723d6aaa5c68)
- [ ] [Set Manager for Department: PUT /departments/{dep_uid}/set-manager/{usr_uid}](https://airqo-engineering.postman.co/workspace/internal~7a8e9fad-2d13-4d27-af27-f1f4e9a114bf/request/12513372-9bcc7862-5faf-4aeb-96fe-94f2ffd599c7)

### Groups:

- [ ] [Get Groups List: GET /groups](https://airqo-engineering.postman.co/workspace/internal~7a8e9fad-2d13-4d27-af27-f1f4e9a114bf/request/12513372-315f9b4f-badf-47f4-817e-f234a9bfd5d9)
- [ ]  [Get Group Information: GET /groups/{grp_uid}](https://airqo-engineering.postman.co/workspace/internal~7a8e9fad-2d13-4d27-af27-f1f4e9a114bf/request/12513372-12b51fb6-949a-42b1-9d5e-abee2d7eaeb5)
- [ ] [Create Group: POST /groups](https://airqo-engineering.postman.co/workspace/internal~7a8e9fad-2d13-4d27-af27-f1f4e9a114bf/request/12513372-a93d1ca8-bb51-4beb-9bdd-b1d538aebe2f)
- [ ] [Update Group: PUT /groups/{grp_uid}](https://airqo-engineering.postman.co/workspace/internal~7a8e9fad-2d13-4d27-af27-f1f4e9a114bf/request/12513372-99a9f2f0-7b59-4052-82de-4e6fb5c8fb78)
- [ ] [Delete Group: DELETE /groups/{grp_uid}](https://airqo-engineering.postman.co/workspace/internal~7a8e9fad-2d13-4d27-af27-f1f4e9a114bf/request/12513372-40d44603-5c78-406c-8664-bad7ee6d9625)
- [ ] [Get Group Members: GET /groups/{grp_uid}/users](https://airqo-engineering.postman.co/workspace/internal~7a8e9fad-2d13-4d27-af27-f1f4e9a114bf/request/12513372-87e04fbc-c6f1-444a-b9c8-64f7cbf7e493)
- [ ] [Get Available Users: GET /groups/{grp_uid}/available-users](https://airqo-engineering.postman.co/workspace/internal~7a8e9fad-2d13-4d27-af27-f1f4e9a114bf/request/12513372-71717d59-144e-4aad-a5d6-4ff217c1d980)
- [ ] [Assign User to Group: POST /groups/{grp_uid}/user](https://airqo-engineering.postman.co/workspace/internal~7a8e9fad-2d13-4d27-af27-f1f4e9a114bf/request/12513372-f09effed-62cb-4e5b-8f3e-b089171d3a56)
- [ ] [Batch assignment of users to a group: POST /groups/batch-users](https://airqo-engineering.postman.co/workspace/internal~7a8e9fad-2d13-4d27-af27-f1f4e9a114bf/request/12513372-fbee3208-1f13-4270-919e-4855d92b77c6)
- [ ] [Unassign User from Group: DELETE /groups/{grp_uid}/unassign-user/{usr_uid}](https://airqo-engineering.postman.co/workspace/internal~7a8e9fad-2d13-4d27-af27-f1f4e9a114bf/request/12513372-27b792d8-0d4c-4003-8035-292dc687c9aa)

[AN-148]: https://airqoteam.atlassian.net/browse/AN-148?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/auth-service/models/Group.js [hunk 1]
```diff
@@ -5,80 +5,53 @@ const validator = require("validator");
 var uniqueValidator = require("mongoose-unique-validator");
 const { logObject, logElement, logText } = require("../utils/log");
 const isEmpty = require("is-empty");
-const createOrganizationUtil = require("../utils/create-organization");
 const { getModelByTenant } = require("../utils/multitenancy");
 const HTTPStatus = require("http-status");
```
Comment (Baalmart) [inline src/auth-service/models/Group.js:10]: Organisation schema renamed to Group to introduce new entity

File: src/auth-service/models/Inquiry.js [hunk 2]
```diff
@@ -25,6 +25,9 @@ const InquirySchema = new mongoose.Schema(
     },
     message: { type: String, required: [true, "message is required"] },
     category: { type: String, required: [true, "category is required"] },
```
Comment (Baalmart) [inline src/auth-service/models/Inquiry.js:27]: adding network as a requirement for inquiries

File: src/auth-service/models/Permission.js [hunk 3]
```diff
@@ -0,0 +1,227 @@
+const mongoose = require("mongoose").set("debug", true);
+const { logObject } = require("../utils/log");
+const isEmpty = require("is-empty");
```
Comment (Baalmart) [inline src/auth-service/models/Permission.js:3]: Adding permissions model

File: src/auth-service/models/Role.js [hunk 4]
```diff
@@ -0,0 +1,255 @@
+const mongoose = require("mongoose").set("debug", true);
+const { logObject, logText } = require("../utils/log");
```
Comment (Baalmart) [inline src/auth-service/models/Role.js:2]: Adding the new Roles model

File: src/auth-service/models/Scope.js [hunk 5]
```diff
@@ -0,0 +1,231 @@
+const mongoose = require("mongoose").set("debug", true);
+const { logObject } = require("@utils/log");
```
Comment (Baalmart) [inline src/auth-service/models/Scope.js:2]: Adding the scope model, should the network ID also be required here? @Codebmk / @123MwanjeMike / @wabinyai

File: src/auth-service/routes/index.js [hunk 6]
```diff
@@ -0,0 +1,7 @@
+const v1 = require("./v1");
```
Comment (Baalmart) [inline src/auth-service/routes/index.js:1]: reorganising our routes

**Review by Codebmk — APPROVED** [review]
@Baalmart great work here. Will the guest user be receiving a user id? During token creation, how is the guest user's token created because token creation requires a userid?

**Comment by Baalmart** [comment]
Thanks @Codebmk for this.

The access token requires one to have an account.

These access tokens are additional authentication mechanisms to the existing JWT. They are not replacing the JWTs.

On that note, a guest user implementation is about to be made where a user without an account receives a JWT token assigned to an "anonymous" user.

I am looking forward to more clarification questions.

**Comment by Codebmk** [comment]
Awesome to hear. Have you already added the client token field as a parameter to the different endpoints? @Baalmart

**Comment by Baalmart** [comment]
Hi @Codebmk that would have to be done at the gateway level. @123MwanjeMike needs to join this conversation asap!

**Comment by Baalmart** [comment]
Also @Codebmk please table this for our next sync. Just realised that things might not be obvious.

With our microservice architecture, things are not very straightforward. But we shall handle. 😀😀

**Comment by Codebmk** [comment]
> Also @Codebmk please table this for our next sync. Just realized that things might not be obvious.
> 
> With our microservice architecture, things are not very straightforward. But we shall handle. grinninggrinning

yeah sure

**Review by 123MwanjeMike — APPROVED** [review]
Thank you @Baalmart

**Pull request merged**